### ISSUE TITLE
Slowness and high memory consumption when operating locally with large databases

### ISSUE THREAD

**Issue by deatinor** [Netflix/metaflow#157]
I am using `metaflow` locally on a machine with 128GB of RAM.

This is a simple version of my code:

    class Flow(FlowSpec):
        
        @step
        def start(self):
            path_list = ...
            self.dataframe = pd.concat([pd.read_pickle(path) for path in path_list])
            self.next(self.end)

        @step
        def end(self):
            pass

Running this in `metaflow` results in `MemoryError` and long running time before crashing (~10 minutes). 
If I run the same commands in a Python script without `metaflow`, everything works and the program is much faster (~4-5 minutes)

The dataset is large, but it becomes intractable with `metaflow`.
 
Do you know why does this happen and if it has to be expected? It is a pity, because I was really enjoying `metaflow`.

Thank you!

**Comment by seeravikiran** [comment]
cc @bergdavidj (for a second opinion)

@deatinor Sorry to hear about your experience. Could you add more details how large these individual dataframe(s) that are being concatenated are? It might help us in debugging your situation better.

Could you also try a quick experiment seeing what happens if you don't store the dataframe in `self.dataframe` but simply in a local variable `dataframe`?
I'm curious to see (if and) how much it affects the performance (memory and latency).

**Comment by marcellovictorino** [comment]
@crk-codaio/ @savingoyal 
It has been a while since this conversation started, but I am now facing a similar issue where there seems to be "memory leakeage"/duplication when assigning values as the class attribute (`self.df`)
The. output of my pandas dataframe.info() shows around 3Gb for ~7 million rows with 100+ columns

But saving artifacts seems to be causing duplication of the data, instead of just pointing to the reference/variable)
This has now become a problem for me when deploying a model using AWS Batch

Am I missing something?

**Comment by romain-intel** [comment]
Yes, there is some duplication of data because artifacts are all pickled and then tarred. Although we do our best to avoid needless duplication, both of those operations require the system to hold the pre-operation input and the post operation output at least for some time.

We are looking to see how we can improve the story around this. We have been talking about this specific type of problem for a while but never got around to implementing something. A workaround for now would be to use the S3 client to directly load/store the large artifact instead of relying on the automagic self. notation. I realize it is not ideal and hopefully this will improve in 2023. Stay tuned.
